File: test-junit5/src/test/java/io/micronaut/test/junit5/MathMockServiceTest.java

package io.micronaut.test.junit5;

import io.micronaut.context.annotation.Requires;
import io.micronaut.core.util.StringUtils;
import io.micronaut.test.annotation.MockBean;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import jakarta.inject.Inject;

import static org.mockito.Mockito.*;

@MicronautTest
@Requires(property = "mockito.test.enabled", defaultValue = StringUtils.FALSE, value = StringUtils.TRUE)
class MathMockServiceTest {

    @Inject
    MathService mathService; // <3>


    @ParameterizedTest
    @CsvSource({"2,4", "3,9"})
    void testComputeNumToSquare(Integer num, Integer square) {

        when(mathService.compute(10))
            .then(invocation -> Long.valueOf(Math.round(Math.pow(num, 2))).intValue());

        final Integer result = mathService.compute(10);

        Assertions.assertEquals(
                square,
                result
        );
        verify(mathService).compute(10); // <4>
    }

    @MockBean(MathServiceImpl.class) // <1>
    MathService mathService() {
        return mock(MathService.class); // <2>
    }

}

File: test-suite-python/src/test/python/micronaut/test/python/MathMockServiceTest.py
from typing import Annotated

from jakarta.inject import Inject, Singleton
from micronaut.context.annotation import Property, Replaces, Requires
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .MathService import MathService
from .MathServiceImpl import MathServiceImpl


@Singleton
@Requires(property="spec.name", value="MathMockServiceTest")  # <1>
@Replaces(MathServiceImpl)  # <2>
class MockMathService(MathService):

    def __init__(self):
        self.result = 0
        self.calls: list[int] = []

    def compute(self, num: int) -> int:
        self.calls.append(num)
        return self.result


@MicronautTest
@Property(name="spec.name", value="MathMockServiceTest")  # <3>
class MathMockServiceTest:

    math_service: Annotated[MathService, Inject]  # <4>

    @Test
    def test_compute_num_to_square(self):
        for num, square in [(2, 4), (3, 9)]:
            self.math_service.calls.clear()
            self.math_service.result = num * num

            result = self.math_service.compute(10)

            assert square == result
            assert [10] == self.math_service.calls  # <5>
